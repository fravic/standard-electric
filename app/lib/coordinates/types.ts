import { z } from "zod";

export const HexCoordinatesSchema = z.object({
  x: z.number(),
  z: z.number(),
});

export type HexCoordinates = z.infer<typeof HexCoordinatesSchema>;

export enum CornerPosition {
  North,
  South,
}

export const CornerCoordinatesSchema = z.object({
  hex: HexCoordinatesSchema,
  position: z.nativeEnum(CornerPosition),
});

export type CornerCoordinates = {
  hex: HexCoordinates;
  position: CornerPosition;
};
