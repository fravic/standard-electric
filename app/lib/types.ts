export type Vertex = readonly [number, number, number];

/**
 * CommodityType enum - defines the types of commodities available
 */
export enum CommodityType {
  COAL = "coal",
  OIL = "oil",
  GAS = "gas",
  URANIUM = "uranium",
}