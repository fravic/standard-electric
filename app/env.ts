import { AnyActorServer } from "actor-kit";
import { GameServer } from "./actor/game.server";
import { Remix } from "./remix.server";

declare module "@remix-run/cloudflare" {
  interface AppLoadContext {
    env: Env;
    userId: string;
    sessionId: string;
    pageSessionId: string;
  }
}

export interface Env {
  GAME: DurableObjectNamespace<GameServer>;
  REMIX: DurableObjectNamespace<Remix>;
  ACTOR_KIT_SECRET: string;
  ACTOR_KIT_HOST: string;
  SESSION_JWT_SECRET: string;
  [key: string]: DurableObjectNamespace<AnyActorServer> | unknown;

  // Required for auth-kit
  AUTH_SECRET: string;
  SENDGRID_API_KEY: string;

  // KV Storage for auth data
  KV_STORAGE: KVNamespace;
}
