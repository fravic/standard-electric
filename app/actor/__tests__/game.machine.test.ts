import { createActor } from "xstate";
import { describe, test, expect, beforeEach } from "bun:test";
import { gameMachine } from "../game.machine";
import { GameEvent, GameInput } from "../game.types";
import { Env } from "../../env";

describe("Game Machine", () => {
  // Create mock objects for testing
  const createMockStorage = () => {
    return {
      get: async () => null,
      put: async () => {},
      delete: async () => {},
      list: async () => ({ keys: [] }),
      getAlarm: async () => null,
      setAlarm: async () => {},
      deleteAlarm: async () => {}
    } as any;
  };

  const createMockEnv = () => {
    return {
      GAME: {},
      REMIX: {},
      ACTOR_KIT_SECRET: "test-secret",
      ACTOR_KIT_HOST: "test-host",
      SESSION_JWT_SECRET: "test-jwt-secret",
      AUTH_SECRET: "test-auth-secret",
      SENDGRID_API_KEY: "test-sendgrid-key",
      KV_STORAGE: {}
    } as Env;
  };

  // Test that the game machine can be created, players can join, and game can start
  test("should allow players to join and start the game", () => {
    // 1. Arrange - Create the game machine and actor
    const mockEnv = createMockEnv();
    const mockStorage = createMockStorage();
    
    const gameActor = createActor(gameMachine, {
      input: {
        id: "test-game",
        caller: { id: "system", type: "system" },
        env: mockEnv,
        storage: mockStorage,
      } as GameInput,
    });

    // Start the actor (initialize the state machine)
    gameActor.start();

    // Initial state should be 'lobby'
    expect(gameActor.getSnapshot().value).toEqual("lobby");

    // 2. Act - Add two players to the game
    // Player 1 joins (becomes host)
    gameActor.send({
      type: "JOIN_GAME",
      name: "Player 1",
      caller: { id: "player-1", type: "client" },
      origin: "client",
      env: mockEnv,
      storage: mockStorage,
    } as unknown as GameEvent);

    // Player 2 joins
    gameActor.send({
      type: "JOIN_GAME",
      name: "Player 2",
      caller: { id: "player-2", type: "client" },
      origin: "client",
      env: mockEnv,
      storage: mockStorage,
    } as unknown as GameEvent);

    // 3. Assert - Check that both players are in the game
    const snapshot = gameActor.getSnapshot();
    const players = snapshot.context.public.players;
    
    expect(Object.keys(players).length).toEqual(2);
    expect(players["player-1"]).toBeDefined();
    expect(players["player-2"]).toBeDefined();
    expect(players["player-1"].name).toEqual("Player 1");
    expect(players["player-2"].name).toEqual("Player 2");
    expect(players["player-1"].isHost).toEqual(true);
    expect(players["player-2"].isHost).toEqual(false);

    // 4. Act - Start the game (only the host can do this)
    gameActor.send({
      type: "START_GAME",
      caller: { id: "player-1", type: "client" },
      origin: "client",
      env: mockEnv,
      storage: mockStorage,
    } as unknown as GameEvent);

    // 5. Assert - Check that the game state has changed to 'auction'
    const finalSnapshot = gameActor.getSnapshot();
    expect(finalSnapshot.value).toEqual({
      auction: "initiatingBid"
    });
    
    // Check that the auction state is initialized
    expect(finalSnapshot.context.public.auction).toBeDefined();
  });
});
