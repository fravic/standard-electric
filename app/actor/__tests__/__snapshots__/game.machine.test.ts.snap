// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Machine should allow players to join, start the game, and perform various actions 1`] = `
{
  "auction": {
    "availableBlueprintIds": [
      "coal_plant_large",
    ],
    "currentBlueprint": null,
    "isPassingAllowed": false,
    "passedPlayerIds": [],
    "purchases": [
      {
        "blueprintId": "coal_plant_small",
        "playerId": "player-1",
        "price": 30,
      },
      {
        "blueprintId": "coal_plant_medium",
        "playerId": "player-2",
        "price": 18,
      },
    ],
  },
  "commodityMarket": {
    "commodities": {
      "coal": {
        "config": {
          "baseExchangeRate": 10,
          "maxExchangeRate": 30,
          "minExchangeRate": 5,
          "priceIncrement": 0.2,
          "tradingFee": 0.3,
          "unitSize": 10,
        },
        "currentExchangeRate": 10.4,
      },
      "gas": {
        "config": {
          "baseExchangeRate": 15,
          "maxExchangeRate": 40,
          "minExchangeRate": 8,
          "priceIncrement": 0.25,
          "tradingFee": 0.3,
          "unitSize": 8,
        },
        "currentExchangeRate": 15,
      },
      "oil": {
        "config": {
          "baseExchangeRate": 20,
          "maxExchangeRate": 50,
          "minExchangeRate": 10,
          "priceIncrement": 0.3,
          "tradingFee": 0.3,
          "unitSize": 5,
        },
        "currentExchangeRate": 20,
      },
      "uranium": {
        "config": {
          "baseExchangeRate": 50,
          "maxExchangeRate": 100,
          "minExchangeRate": 25,
          "priceIncrement": 0.5,
          "tradingFee": 0.3,
          "unitSize": 1,
        },
        "currentExchangeRate": 50,
      },
    },
  },
  "entitiesById": {
    "ENTITY_ID_0": {
      "blueprint": {
        "allowedPosition": "corner",
        "components": {
          "connections": {
            "connectedToIds": [],
          },
          "renderable": {
            "renderableComponentName": "PowerPole",
          },
        },
        "name": "Power pole",
      },
      "id": "ENTITY_ID_0",
      "name": "Power pole blueprint",
      "owner": {
        "playerId": "player-1",
      },
    },
    "ENTITY_ID_1": {
      "blueprint": {
        "allowedPosition": "corner",
        "components": {
          "connections": {
            "connectedToIds": [],
          },
          "renderable": {
            "renderableComponentName": "PowerPole",
          },
        },
        "name": "Power pole",
      },
      "id": "ENTITY_ID_1",
      "name": "Power pole blueprint",
      "owner": {
        "playerId": "player-2",
      },
    },
    "ENTITY_ID_2": {
      "blueprint": {
        "allowedPosition": "hex",
        "components": {
          "fuelRequirement": {
            "fuelConsumptionPerKWh": 0.1,
            "fuelType": "coal",
          },
          "fuelStorage": {
            "currentFuelStorage": 0,
            "fuelType": "coal",
            "maxFuelStorage": 1000,
          },
          "powerGeneration": {
            "powerGenerationKW": 1000,
            "pricePerKWh": 0.1,
          },
          "renderable": {
            "renderableComponentName": "PowerPlant",
          },
        },
        "name": "Small Coal Plant",
      },
      "cost": {
        "amount": 10,
      },
      "id": "ENTITY_ID_2",
      "name": "Small Coal Plant",
      "owner": {
        "playerId": "player-1",
      },
    },
    "ENTITY_ID_3": {
      "blueprint": {
        "allowedPosition": "hex",
        "components": {
          "fuelRequirement": {
            "fuelConsumptionPerKWh": 0.09,
            "fuelType": "coal",
          },
          "fuelStorage": {
            "currentFuelStorage": 0,
            "fuelType": "coal",
            "maxFuelStorage": 1800,
          },
          "powerGeneration": {
            "powerGenerationKW": 2000,
            "pricePerKWh": 0.1,
          },
          "renderable": {
            "renderableComponentName": "PowerPlant",
          },
          "requiredRegion": {
            "requiredRegionName": "California",
          },
        },
        "name": "Medium Coal Plant",
      },
      "cost": {
        "amount": 18,
      },
      "id": "ENTITY_ID_3",
      "name": "Medium Coal Plant",
      "owner": {
        "playerId": "player-2",
      },
    },
    "ENTITY_ID_4": {
      "blueprint": {
        "allowedPosition": "hex",
        "components": {
          "fuelRequirement": {
            "fuelConsumptionPerKWh": 0.08,
            "fuelType": "coal",
          },
          "fuelStorage": {
            "currentFuelStorage": 0,
            "fuelType": "coal",
            "maxFuelStorage": 2500,
          },
          "powerGeneration": {
            "powerGenerationKW": 3000,
            "pricePerKWh": 0.1,
          },
          "renderable": {
            "renderableComponentName": "PowerPlant",
          },
          "requiredRegion": {
            "requiredRegionName": "Texas",
          },
        },
        "name": "Large Coal Plant",
      },
      "cost": {
        "amount": 25,
      },
      "id": "ENTITY_ID_4",
      "name": "Large Coal Plant",
    },
    "ENTITY_ID_5": {
      "connections": undefined,
      "fuelRequirement": {
        "fuelConsumptionPerKWh": 0.1,
        "fuelType": "coal",
      },
      "fuelStorage": {
        "currentFuelStorage": 10,
        "fuelType": "coal",
        "maxFuelStorage": 1000,
      },
      "hexPosition": {
        "coordinates": {
          "x": 3,
          "z": 11,
        },
      },
      "id": "ENTITY_ID_5",
      "name": "Small Coal Plant",
      "owner": {
        "playerId": "player-1",
      },
      "powerGeneration": {
        "powerGenerationKW": 1000,
        "pricePerKWh": 0.1,
      },
      "renderable": {
        "renderableComponentName": "PowerPlant",
      },
    },
    "ENTITY_ID_6": {
      "connections": undefined,
      "fuelRequirement": {
        "fuelConsumptionPerKWh": 0.09,
        "fuelType": "coal",
      },
      "fuelStorage": {
        "currentFuelStorage": 10,
        "fuelType": "coal",
        "maxFuelStorage": 1800,
      },
      "hexPosition": {
        "coordinates": {
          "x": 4,
          "z": 14,
        },
      },
      "id": "ENTITY_ID_6",
      "name": "Medium Coal Plant",
      "owner": {
        "playerId": "player-2",
      },
      "powerGeneration": {
        "powerGenerationKW": 2000,
        "pricePerKWh": 0.1,
      },
      "renderable": {
        "renderableComponentName": "PowerPlant",
      },
      "requiredRegion": {
        "requiredRegionName": "California",
      },
    },
    "ENTITY_ID_7": {
      "connections": {
        "connectedToIds": [],
      },
      "cornerPosition": {
        "cornerCoordinates": {
          "hex": {
            "x": 3,
            "z": 11,
          },
          "position": 0,
        },
      },
      "id": "ENTITY_ID_7",
      "name": "Power pole",
      "owner": {
        "playerId": "player-1",
      },
      "renderable": {
        "renderableComponentName": "PowerPole",
      },
    },
    "ENTITY_ID_8": {
      "connections": {
        "connectedToIds": [
          "ENTITY_ID_7",
        ],
      },
      "cornerPosition": {
        "cornerCoordinates": {
          "hex": {
            "x": 4,
            "z": 10,
          },
          "position": 1,
        },
      },
      "id": "ENTITY_ID_8",
      "name": "Power pole",
      "owner": {
        "playerId": "player-1",
      },
      "renderable": {
        "renderableComponentName": "PowerPole",
      },
    },
  },
  "id": "test-game",
  "players": {
    "player-1": {
      "isHost": true,
      "money": 41.42,
      "name": "Player 1",
      "number": 1,
      "powerSoldKWh": 0,
    },
    "player-2": {
      "isHost": false,
      "money": 50.48,
      "name": "Player 2",
      "number": 2,
      "powerSoldKWh": 0,
    },
  },
  "randomSeed": 12345,
  "time": {
    "isPaused": true,
    "totalTicks": 5,
  },
}
`;

exports[`Game Machine should allow players to join, start the game, and perform various actions 2`] = `
{
  "entitiesById": {
    "ENTITY_ID_0": {
      "hexPosition": {
        "coordinates": {
          "x": 3,
          "z": 11,
        },
      },
      "id": "ENTITY_ID_0",
      "name": "Survey for 3,11",
      "owner": {
        "playerId": "player-1",
      },
      "surveyResult": {
        "isComplete": true,
        "resource": undefined,
        "surveyStartTick": 0,
      },
    },
  },
}
`;

exports[`Game Machine should allow players to join, start the game, and perform various actions 3`] = `
{
  "entitiesById": {
    "ENTITY_ID_0": {
      "hexPosition": {
        "coordinates": {
          "x": 4,
          "z": 14,
        },
      },
      "id": "ENTITY_ID_0",
      "name": "Survey for 4,14",
      "owner": {
        "playerId": "player-2",
      },
      "surveyResult": {
        "isComplete": true,
        "resource": undefined,
        "surveyStartTick": 0,
      },
    },
  },
}
`;
